fms/utils/generation.py

```diff
@@ -53,30 +150,40 @@ def generate(
         use_cache: requires that the model accept use_cache and
             past_key_value_states args in forward method.
     """
-    batched = False
     if num_beams != 1:
         raise NotImplementedError("generate() does yet not support beam search")
-    if type(input_ids) == torch.Tensor:
-        if input_ids.dim() != 1:
-            batched = True
-    else:
-        raise RuntimeError("generate() requires a tensor of token ids as the prefix")
 
-    if not batched:
-        input_ids = input_ids.unsqueeze(0)
+    kwargs: MutableMapping[str, Any] = dict()
+    # if the inputs are a tensor, we assume they are all non-pad ids and include entire context length
+    if isinstance(input_ids, torch.Tensor):
+        is_batch = len(input_ids.shape) > 1
+        # our model requires batch dimension
+        if not is_batch:
+            input_ids = input_ids.unsqueeze(0)
+    # if the inputs are a list, they may be made up of differently sized tensors
+    # in the case where the tensors are of different sizes, proper position ids and pads will be created
+    elif isinstance(input_ids, List):
+        is_batch = len(input_ids) > 1
+        input_ids, kwargs = __prepare_list_input(input_ids, model)
+    else:
+        raise TypeError("input_ids must be one of Tensor or List")
 
     eos_found = torch.zeros(
         input_ids.shape[0], dtype=torch.bool, device=input_ids.device
     )
 
     result = input_ids
     next_input = input_ids
-    kwargs: MutableMapping[str, Any] = dict()
     kwargs["past_key_value_states"] = None
     kwargs["use_cache"] = use_cache
-
-    for _ in range(max_new_tokens):
+    for i in range(max_new_tokens):
         input_ids = next_input[:, -max_seq_len:]
+
+        # prepare any padding keyword arguments
+        # iteration 0 is the prefill step (cache has not been filled yet), so no need to extend the mask/position_ids
+        if i > 0:
+            __update_padding_kwargs(use_cache, kwargs)
```

Comment:
nit: I think this would be clearer in a functional style like `kwargs = _update_padding_kwargs(kwargs, use_cache)`

Reply:
I agree! Updated in latest commit